File: Source/RTSPPlayerClient/Base/Serialization/InterprocessSerializer.cpp
/// \file InterprocessSerializer.cpp
/// \brief Contains definitions of classes and functions for processing
/// interprocess data.
/// \bug No known bugs.

#include "InterprocessSerializer.hpp"

namespace {

    /// Encodes a regular byte array to Base64 format.
    /// \details Encodes a byte array to Base64 format based on RFC 4648.
    /// \param[in]  array   Regular byte array.
    /// \return Encoded Base64 byte array.
    QByteArray encodeBase64(const QByteArray& array) {
        return array.toBase64(QByteArray::Base64Encoding |
                              QByteArray::OmitTrailingEquals);
    }

    /// Decodes a Base64 byte array to a regular byte array.
    /// \details Decodes a byte array from Base64 format based on RFC 4648.
    /// \param[in]  array   Base64 byte array.
    /// \return Decoded regular byte array.
    QByteArray decodeBase64(const QByteArray& array) {
        return QByteArray::fromBase64(array, QByteArray::Base64Encoding |
                                      QByteArray::IgnoreBase64DecodingErrors);
    }
}

/// A namespace that contains common classes and functions for data
/// serialization.
namespace Common::Serialization {

    /// Serializes the interprocess frame into a byte array.
    /// \details Delegates serialization to an overloaded function.
    /// \param[in]  frame   Interprocess frame.
    /// \return Byte array.
    QByteArray InterprocessSerializer::serialize(
        const InterprocessFrame& frame) const {

        QByteArray array;
        serialize(frame, array);
        return array;
    }

    /// Serializes the interprocess frame into a byte array.
    /// \details Concatenates the key-value dictionary into a byte array.
    /// \param[in]  frame   Interprocess frame.
    /// \param[out] array   Byte array.
    void InterprocessSerializer::serialize(const InterprocessFrame& frame,
                                           QByteArray& array) const {

        auto begin = frame.parameterDictionary.cbegin();
        auto end = frame.parameterDictionary.cend();

        for (auto it = begin; it != end; ++it)
            array
                    += encodeBase64(it.key().toUtf8())
                    + '='
                    + encodeBase64(it.value().toUtf8())
                    + ' ';

        array += '\n';
    }

    /// Deserializes the byte array into an interprocess frame.
    /// \details Delegates deserialization to an overloaded function.
    /// \param[in]  array   Byte array.
    /// \return Interprocess frame.
    InterprocessFrame InterprocessSerializer::deserialize(
        const QByteArray& array) {

        InterprocessFrame frame;
        deserialize(array, frame);
        return frame;
    }

    /// Deserializes the data buffer into an interprocess frame.
    /// \details Delegates deserialization to an overloaded function.
    /// \param[in]  data    Data buffer.
    /// \param[in]  size    Buffer size.
    /// \return Interprocess frame.
    InterprocessFrame InterprocessSerializer::deserialize(
        const char* data, int size) {

        InterprocessFrame frame;
        deserialize(data, size, frame);
        return frame;
    }

    /// Deserializes the byte array into an interprocess frame.
    /// \details Parses the byte array into a key-value dictionary.
    /// \param[in]  array   Byte array.
    /// \param[out] frame   Interprocess frame.
    void InterprocessSerializer::deserialize(const QByteArray& array,
                                             InterprocessFrame& frame) {

        auto pairs = array.toLower().split(' ');

        for (const auto& pair : pairs) {
            auto index = pair.indexOf('=');
            if (index <= 0 || index >= pair.size() - 1) continue;

            auto key = decodeBase64(pair.left(index).trimmed());
            auto value = decodeBase64(pair.mid(index + 1).trimmed());

            if (!key.isEmpty())
                frame.parameterDictionary.insert(key, value);
        }
    }

    /// Deserializes the data buffer into an interprocess frame.
    /// \details Delegates deserialization to an overloaded function.
    /// \param[in]  data    Data buffer.
    /// \param[in]  size    Buffer size.
    /// \param[out] frame   Interprocess frame.
    void InterprocessSerializer::deserialize(const char* data, int size,
                                             InterprocessFrame& frame) {

        deserialize(QByteArray::fromRawData(data, size), frame);
    }
}

File: Source/RTSPPlayerServer/RTSPPlayerServer.Service/Serializers/Interprocess/InterprocessSerializer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using RTSPPlayerServer.Service.Base.Extensions;

namespace RTSPPlayerServer.Service.Serializers.Interprocess
{
    /// <summary>
    /// A class that provides an interprocess serializer implementation.
    /// </summary>
    public class InterprocessSerializer : IInterprocessSerializer
    {
        /// <summary>
        /// A padding character in Base64 strings.
        /// </summary>
        private static char PaddingCharacter => '=';

        /// <summary>
        /// A character to separate keys and values in pairs.
        /// </summary>
        private static char PairCharacter => '=';

        /// <summary>
        /// A character to separate data records.
        /// </summary>
        private static char DelimiterCharacter => ' ';

        /// <summary>
        /// Serializes the interprocess frame into a string message.
        /// </summary>
        /// <param name="interprocessFrame">Interprocess frame.</param>
        /// <returns>Serialized message.</returns>
        public string Serialize(InterprocessFrame interprocessFrame)
        {
            return interprocessFrame.ParameterDictionary.Aggregate(
                string.Empty,
                (current, parameter) =>
                    current
                    + DelimiterCharacter
                    + $"{EncodeBase64(parameter.Key.ToLowerInvariant())}"
                    + PairCharacter
                    + $"{EncodeBase64(parameter.Value)}"

            ).ToLower();
        }

        /// <summary>
        /// Deserializes the string message into an interprocess frame.
        /// </summary>
        /// <param name="message">Serialized message.</param>
        /// <returns>Interprocess frame.</returns>
        public InterprocessFrame Deserialize(string message) =>
            new InterprocessFrame(ParseMessage(message));

        /// <summary>
        /// Parses the string message into the key-value parameter dictionary.
        /// </summary>
        /// <param name="message">Serialized message.</param>
        /// <returns>Parameter dictionary.</returns>
        private static Dictionary<string, string> ParseMessage(string message)
        {
            var dictionary = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
            var pairs = message.Split(DelimiterCharacter, StringSplitOptions.RemoveEmptyEntries);
            if (!pairs.Any()) return dictionary;

            foreach (var pair in pairs)
            {
                var index = pair.IndexOf(PairCharacter);
                if (index <= 0 || index >= pair.Length - 1) continue;

                var key = DecodeBase64(pair.Substring(0, index).Trim());
                var value = DecodeBase64(pair.Substring(index + 1).Trim());
                if (string.IsNullOrEmpty(key) || string.IsNullOrEmpty(value)) continue;

                dictionary.AddOrUpdate(key.ToLowerInvariant(), value);
            }

            return dictionary;
        }

        /// <summary>
        /// Encodes a regular string to Base64 format.
        /// </summary>
        /// <param name="data">Regular string.</param>
        /// <returns>Encoded Base64 string.</returns>
        private static string? EncodeBase64(string? data)
        {
            if (data == null) return null;

            var bytes = Encoding.UTF8.GetBytes(data);
            return Convert.ToBase64String(bytes).TrimEnd(PaddingCharacter);
        }

        /// <summary>
        /// Decodes a Base64 string to a regular string.
        /// </summary>
        /// <param name="data">Base64 string.</param>
        /// <returns>Decoded regular string.</returns>
        private static string? DecodeBase64(string? data)
        {
            if (data == null) return null;

            var length = (data.Length % 4) switch
            {
                2 => 2,
                3 => 1,
                _ => 0
            } + data.Length;

            var bytes = Convert.FromBase64String(data.PadRight(length, PaddingCharacter));
            return Encoding.UTF8.GetString(bytes);
        }
    }
}
